Dependency Scanning Library

artemcm opened the pull request:
This PR introduces a dependency scanning tool/library that is meant to be used by `libSwiftDriver`, replacing its current calls to `swift-frontend -scan-dependencies`. 

This library is designed to be used for dependency scanning across multiple, different modules, with a common `ModuleDependencyCache`, in order to re-use computed information about already-scanned modules.

This PR:
- Adds a C++ interface for a tool that answers dependency-scanning queries (`DependencyScanningTool`).
- Adds a unittest harness to invoke the `DependencyScanningTool` library.
- Factors ModuleDependenciesCache out of a `CompilerInstance` into callers of `scanDependencies`. It is now the responsibility of the `scanDependencies` code to instantiate (and share) the cache. e.g. `FrontendTool` instantiates a new cache per `-scan-dependencies` invocation, and the `DependencyScanningTool` keeps one shared cache across its lifetime. (Effectively un-doing https://github.com/apple/swift/pull/34646)
- Deprecates and removes `-scan-clang-dependencies`. Since it was introduced, its use-case has been entirely subsumed by batch scanning.
- Adds a C API layer for dependency scanning tool, scanning actions, and scan results. (`include/swift-c/DependencyScan`)
- Refactors dependency scanner to return an in-memory format as its result output. Specifically, code in `ScanDependencies.cpp` is split into two functional groups:
    - Scan execution code that performs the mechanics of the scan and produces an in-memory result
    - Dependency scanner entry-points used when the scanning action is invoked as a `swift-frontend` execution mode
    - Adds the aforementioned in-memory dependency scanning result in `DependencyScan.h`, modelled after the `InterModuleDependencyGraph` representation in swift-driver

Resolves rdar://71209248
Resolves rdar://71209311

artemcm committed:
[Dependency Scanning] Add a C++ itnerface for a tool that answers dependency-scanning queries.

artemcm committed:
[Dependency Scanning] Add a unittest harness to invoke the dependency scanning tool C++ library

artemcm committed:
[Dependency Scanning] Factor ModuleDependenciesCache out of a CompilerInstance into callers of scanDependencies
It is now the responsibility of the scanDependencies code to instantiate (and share) the cache.
e.g. FrontendTool instantiates a new cache per `-scan-dependencies` invocation, and the DependencyScanningTool keeps one shared cache across its lifetime.

artemcm committed:
[Dependency Scanning] Add DependencyScanningTool entry-point for batch scanning

artemcm committed:
[Dependency Scanning] Deprecate and remove `-scan-clang-dependencies`
Since it was introduced, its use-case has been entirely subsumed by batch scanning.

artemcm commented on lib/SwiftScan/CMakeLists.txt:
```diff
@@ -0,0 +1,27 @@
+    swiftPrintAsObjC
+    swiftSerialization
+    swiftSIL
+    swiftSILGen
```
We should be able to prune this list quite a bit, I think.

nkcsgexi commented on lib/SwiftScan/ScanDependencies.cpp:
```diff
@@ -27,6 +26,7 @@
 #include "swift/Frontend/FrontendOptions.h"
 #include "swift/Frontend/ModuleInterfaceLoader.h"
 #include "swift/Strings.h"
+#include "swift/SwiftScan/ScanDependencies.h"
```
Nitpick: rename dir `SwiftScan` to `DependenciesScan`

artemcm committed:
Rename `SwiftScan` directory into `DependencyScan`

artemcm committed:
[Dependency Scanning] Factor out import prescan into separate utility

artemcm committed:
[Dependency Scanning] Refactor dependency scanner to return an in-memory format as a result
This commit refactors ScanDependencies.cpp to split the functionality into two functional groups:
- Scan execution code that performs the mechanics of the scan and produces an in-memory result
- Dependency scanner entry-points used when the scanning action is invoked as a `swift-frontend` execution mode
This commit also adds the aforementioned in-memory dependency scanning result in `FullDependencies.h`, modeled after the InterModuleDependencyGraph representation in swift-driver

artemcm commented:
@swift-ci please smoke test

artemcm commented:
@nkcsgexi, curious to hear your thoughts on the refactor in `ScanDependencies.h` and `ScanDependencies.cpp`.

nkcsgexi commented:
They look good to me for now. Can we first define the C header defining those APIs libSwiftDriver need for scanning and refactor the C++ implementation accordingly?

artemcm commented:
> They look good to me for now. Can we first define the C header defining those APIs libSwiftDriver need for scanning and refactor the C++ implementation accordingly?

I like this idea, yeah, I'll do that.

artemcm committed:
[Dependency Scanning] Add a C API layer for dependency scanning tool, scanning actions, and scan results.
Adds a C API layer consisting of:
- Data structures used to represent in-memory result of dependency scanning
- Opaque dependency scanner tool (C wrapper for `DependencyScanningTool`)

Refactors `ScanDependencies.cpp` to produce dependency scanning result in the form of the above binary format.

artemcm committed:
[Dependency Scanning] Add entry-point to DependencyScanningTool and the corresponding C API for an import prescan operation

artemcm commented:
@akyrtzi @benlangmuir, 
I would be grateful to get your comments on the C API, among other things, mostly contained in:
`include/swift-c/DependencyScan`

artemcm commented:
Likewise, @nkcsgexi.

nkcsgexi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,214 @@
+
+  /// A flag to indicate whether or not this module is a framework.
+  bool is_framework;
+} depscan_swift_textual_details_t;
```
Shouldn't we treat `depscan_swift_textual_details_t` as an opaque type and add C APIs to retrieve each field? This allows us to modify the actual contents in the details. I guess this also applies to all other `details` types in this header.

artemcm replied:
Yes, will switch to this, thank you.

artemcm replied:
I made this change. There is now an added file:
`include/swift/DependencyScan/DependencyScanImpl.h`, which contains the implementation of these data structures which we are free to modify, while the public C interface uses an opaque container for details with APIs to query fields. 
The implementation of the C API has been factored out into `lib/DependencyScan/DependencyScanImpl.cpp`.

artemcm committed:
[Dependency Scanning] Use an opaque type for dependency module details with API query functions for individual fields

artemcm committed:
[Dependency Scanning] Add entry-point to DependencyScanningTool and the corresponding C API for a batch scan operation

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,228 @@
+
+  /// Specific details of a particular kind of module.
+  ds_module_details_t details;
+} ds_dependency_info_t;
```
I would recommend making `ds_dependency_info_t` opaque with accessors for all the fields you want to expose.  It seems likely that you might some day want to add more data here.

artemcm replied:
Turned into an opaque type, thank you.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,228 @@
+
+  /// The complete list of modules discovered
+  ds_dependency_set_t *module_set;
+} ds_dependency_result_t;
```
It's less clear to me here, but this might also be worth making opaque.  Basically I would make every struct opaque unless either
a) there is some really strong reason it will never change (e.g. `ds_dependency_set_t` is just an array, so seems pretty safe)
b) field access is performance critical and you have measured that this helps

artemcm replied:
Turned into an opaque type as well.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,228 @@
+  ds_string_t module_name;
+  ds_string_t arguments;
+  bool is_swift;
+} ds_batch_scan_entry_t;
```
Opaque?

artemcm replied:
Turned into an opaque type as well.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,228 @@
+
+DEPSCAN_PUBLIC ds_dependency_result_t *
+ds_scan_dependencies(ds_scanner_t *scanner, const char *working_directory,
+                     int argc, const char *const *argv);
```
How confident are you that you never need more parameters to this function (and the two that follow)?  To allow addition, you could create an opaque "request" with setter functions for the things you need.  Or as a middle ground, you could expose an extra `void *` parameter that should be set to `nullptr` for now, but it leaves you room to add extended parameters in the future.

artemcm replied:
Added an opaque `invocation` context type with setters/getters for its fields and a creation method. Thank you.

akyrtzi commented on include/swift/DependencyScan/DependencyScanImpl.h:
```diff
@@ -0,0 +1,126 @@
+#include "swift-c/DependencyScan/DependencyScan.h"
+#include "swift/DependencyScan/DependencyScanningTool.h"
+
+DEPSCAN_BEGIN_DECLS
```
Use of these macros here should not be necessary since this header is intended to be consumed by C++ code.

artemcm replied:
Removed, thanks for the catch.

akyrtzi commented on include/swift/DependencyScan/DependencyScanImpl.h:
```diff
@@ -0,0 +1,126 @@
+    ds_swift_placeholder_details_t swift_placeholder_details;
+    ds_clang_details_t clang_details;
+  };
+} ds_impl_module_details_t;
```
Since you already have the underlying types of the exported opaque types as C structs (instead of C++) you could make things a bit more convenient on the C++ side by exporting the types as pointers to structs, like this:
```
// in the exported C header
typedef struct ds_module_details_s *ds_module_details_t;

[...]

// in the implementation
struct ds_module_details_s {
  ds_dependency_info_kind_t kind;
  union {..}
};
```
The benefit of this is you have C++ type safety and you don't need a bunch of 'wrap/unwrap' functions to call.
Note that this doesn't make a difference for the Swift side, the pointers will still be opaque when imported in Swift.

artemcm replied:
This is indeed a lot cleaner and the safety is nice to have. Thanks for pointing out this technique.

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,228 @@
+
+//=== Swift Textual Module Details query APIs -----------------------------===//
+DEPSCAN_PUBLIC ds_string_t
+ds_get_swift_textual_detail_module_interface_path(ds_module_details_t details);
```
I'd suggest to try to stick to this naming convention for the functions: \<prefix>\_\<object>\_\<action>
So in this case: `ds_swift_textual_detail_get_module_interface_path`

artemcm replied:
Made all API consistently follow this naming scheme, thank you.

akyrtzi commented:
Some additional things for your consideration:
* Adding an [exports list file](https://github.com/apple/swift/blob/main/tools/libSwiftSyntaxParser/libSwiftSyntaxParser.exports) would be general goodness. It is what will give you a linker error if an exported function is not implemented (e.g. when some type or function changes and the function implementation doesn't match the declaration) and can help to strip the library of unused code and make it smaller.
* `ds` as prefix is too short. Consider that you are essentially adding symbols to the global namespace of a client's process (any client, not necessarily something we control), and the prefix is the "defense" against unintentional conflicts with other libraries. I'd suggest to use a more discriminating prefix, like `swiftds_` or `swiftscan_`
* Consider using `_Nonnull` and `_Nullable` to make the APIs more ergonomic for Swift
* I'd recommend to base your testing on top of the C API. It will be a bit less convenient than testing using the C++ APIs but you will have piece of mind that you are really testing what the client is actually using.

artemcm committed:
[Dependency Scanning] Change C API prefix to: `swiftscan` from `ds` and adopt query method naming scheme of:
`<prefix>_<object>_<action>`

artemcm committed:
[Dependency Scanning] Make a module's dependency info an opaque structure

artemcm committed:
[Dependency Scanning] Make the overall dependency scan result (inter-module dependency graph) an opaque structure

artemcm committed:
[Dependency Scanning] Make the batch scan entry container an opaque type

artemcm committed:
[Dependency Scanning] Make the prescan dependency scan result (import set) an opaque structure

artemcm committed:
[Dependency Scanning] Make scanner invocation context (argc, argv, working_dir) an opaque type

benlangmuir commented on include/swift-c/DependencyScan/DSString.h:
```diff
@@ -10,41 +10,41 @@
   const void *data;
   unsigned private_flags;
-} ds_string_t;
+} swiftscan_string_t;
```
@akyrtzi do you know why CXString has this representation?  I was surprised it's not just a pointer.

artemcm replied:
The only use of the flags seems to be to distinguishing who manages this pointer, whether it was malloc'd locally or points to unmanaged memory.

akyrtzi replied:
I personally dislike `CXString`, I'd suggest to either use something like [`indexstore_string_ref_t`](https://github.com/apple/llvm-project/blob/apple/main/clang/include/indexstore/indexstore.h#L126) and be clear what is the lifetime of the returned string (e.g. "same lifetime as the object the string came from") or malloc and return a `char *` and require the caller to `free()` it.

akyrtzi replied:
> be clear what is the lifetime of the returned string (e.g. "same lifetime as the object the string came from")

Note that on the Swift code you are mainly just going to copy it into a Swift string as soon as you call it, so the lifetime aspect it won't make things complicated (it will even simplify things since you won't have to call dispose on it).

artemcm replied:
I changed the string implementation to one that resembles `index store_string_ref_t` and its lifetime is now tied to the object the string comes from. Clients need not worry about freeing string memory, it will be freed automatically when the owning object is disposed of.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -25,204 +25,230 @@
+swiftscan_clang_detail_get_module_map_path(swiftscan_module_details_t details);
+
+SWIFTSCAN_PUBLIC swiftscan_string_t
+swiftscan_clang_detail_get_context_hash(swiftscan_module_details_t details);
```
These accessors that return `swiftscan_string_t` or `swiftscan_string_set_t *` are not clear about ownership.  I see you have `swiftscan_string_dispose` and `swiftscan_string_set_dispose` methods, but the implementation looks like they are returned _unowned_.  If not every API is consistent about whether the caller needs to dispose of the result, I recommend putting it in the doc comment for the accessors.

artemcm replied:
Resolved by tying the lifetime of the string to the object it is associated with. `swiftscan_string_dispose` and `swiftscan_string_set_dispose` are no longer part of the public API.

CodaFi replied:
In general, [I propose we adopt Foundation-style ownership signals](https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-CJBEJBHH) here

artemcm replied:
> In general, [I propose we adopt Foundation-style ownership signals](https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-CJBEJBHH) here

@CodaFi, thank you taking a look at this PR and proposing this convention. 
The API now consistently follows this naming scheme, providing a `create` and `dispose` functions for all objects that the library's clients may take ownership of.

artemcm committed:
[Dependency Scanning] Export opaque types as pointers to C structs instead of `void*` for better type-safety and cleaner code

artemcm committed:
[Dependency Scanning] Add setters for the scanner invocation context and a creation method to return an isntance of the opaque invocation

artemcm committed:
[Dependency Scanning] Factor the shared library libSwiftScan into `tools` and add exports file.
This library now relires on a static compiler library called `swiftDependencyScan`, which is also common to being used by `swift-frontend` for its dependency scanner invocations.

artemcm committed:
[Dependency Scanning] Refactor string implementation to always malloc and tie lifetime to owning object.

artemcm commented:
> * Adding an [exports list file](https://github.com/apple/swift/blob/main/tools/libSwiftSyntaxParser/libSwiftSyntaxParser.exports) would be general goodness. It is what will give you a linker error if an exported function is not implemented (e.g. when some type or function changes and the function implementation doesn't match the declaration) and can help to strip the library of unused code and make it smaller.

Already caught an error by doing this, thank you @akyrtzi!

artemcm committed:
[Dependency Scanning] Fix modulemap module name for the _InternalSwiftScan module.

CodaFi commented on tools/libSwiftScan/libSwiftScan.exports:
```diff
@@ -0,0 +1,51 @@
+swiftscan_dependency_result_get_main_module_name
```
You may have some entertaining conflicts with [this](https://man7.org/linux/man-pages/man5/exports.5.html).

CodaFi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,301 @@
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef SWIFT_C_DEPENDENCY_SCAN_H
```
`"llvm/Support/CBindingWrapping.h"` provides some utilities that might help things here.

artemcm replied:
Thanks!

artemcm committed:
[Dependency Scanning] Fix the DependencyScan unit test harness to reflect C API changes.

artemcm committed:
[Dependency Scanning] Establish object ownership convention by using a Foundation-style ownership signals naming scheme across the API surface.

Foundation-style ownership signals documentation:
https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFMemoryMgmt/Concepts/Ownership.html#//apple_ref/doc/uid/20001148-CJBEJBHH

artemcm committed:
[Dependency Scanning] Use DEFINE_SIMPLE_CONVERSION_FUNCTIONS macro instead of defining own wrap/unwrap functions for the opaque scanner type.

artemcm committed:
[Dependency Scanning] Bugfix: check for null C string before checking if it is `empty`

artemcm commented:
@swift-ci please test

artemcm commented:
@nkcsgexi @akyrtzi @benlangmuir, I think the feedback above is addressed in the current revision. 
This is a ping for a re-review, at your convenience.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,339 @@
+typedef struct {
+  swiftscan_string_ref_t *strings;
+  unsigned count;
+} swiftscan_string_set_t;
```
Nit: maybe use `size_t` for array lengths in ABI?

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,339 @@
+
+/// Full Dependency Graph (Result)
+typedef struct {
+  int count;
```
Nit: maybe use `size_t` for array lengths in ABI?  Also, your other arrays mostly put the pointer first.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,339 @@
+typedef struct swiftscan_batch_scan_entry_s *swiftscan_batch_scan_entry_t;
+
+typedef struct {
+  int count;
```
Nit: maybe use `size_t` for array lengths in ABI?  Also, your other arrays mostly put the pointer first.

benlangmuir commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,339 @@
+} swiftscan_batch_scan_input_t;
+
+typedef struct {
+  int count;
```
Nit: maybe use `size_t` for array lengths in ABI?  Also, your other arrays mostly put the pointer first.

artemcm committed:
[Dependency Scanning] Use `size_t` for array sizes and re-arrange array struct type fields to be consistent.

artemcm commented:
@swift-ci please test

artemcm committed:
[Dependency Scanning] Cache compiler instance/module depenency cache used for batched versioned-PCM scans.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Linux platform

artemcm commented:
@swift-ci please test Windows platform

akyrtzi commented on include/swift-c/DependencyScan/DependencyScan.h:
```diff
@@ -0,0 +1,339 @@
+
+/// Retrieve the character data associated with the given string.
+SWIFTSCAN_PUBLIC const char *
+swiftscan_get_C_string(swiftscan_string_ref_t string);
```
Is it guaranteed that a `swiftscan_string_ref_t` will always point to a null-terminated buffer?
If not, I'd suggest to avoid having this function and on the Swift side create a Swift string using pointer+size.
It will be safer since you won't have to worry about always providing null-terminated strings.

artemcm replied:
It is not guaranteed, indeed.
Removed the function and will instead build a client-side utility as you describe. 
Thanks for the suggestion!

artemcm committed:
[Dependency Scanning] Remove the C string getter in favour of client-side string construction
(Using string length provided by the type)

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test macOS Platform

artemcm commented:
Ah, it looks like our Windows C++ compiler does not support [designated initializers](https://en.cppreference.com/w/cpp/language/aggregate_initialization#Designated_initializers) yet.

artemcm committed:
Do not use Designated Initializers as they are not yet supported by all of our build compilers.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test

nkcsgexi reviewed (approved):
I'm so excited to see this being implemented👍

artemcm committed:
[Dependency Scanning] Skip over source-code (not interface) modules when collecting cross import overlays

artemcm committed:
[Dependency Scanning] Update cached batch scan compiler instances with up-to-date search paths on successive scanner invocations
To ensure they do not get stuck with stale search paths when scanning successive targets.

artemcm committed:
[Dependency Scanning] Ensure INITIALIZE_LLVM() is invoked on scanner creation to populate LLVM's target data store.

artemcm commented:
@swift-ci please test Linux platform

artemcm committed:
Change file generation in the DependencyScan.ModuleDeps unittest to hopefully work better on windows.

artemcm commented:
@swift-ci please test

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Windows platform

artemcm commented:
@swift-ci please test Linux platform

artemcm commented:
@swift-ci please test macOS platform

artemcm commented:
@swift-ci please test macOS platform

artemcm commented:
@swift-ci please test macOS platform

nkcsgexi commented:
🎉
